### PR TITLE
[SPARK-27846][CORE] Eagerly compute Configuration.properties in sc.hadoopConfiguration

### PR DESCRIPTION
**Pull request by JoshRosen**
## What changes were proposed in this pull request?

Hadoop `Configuration` has an internal `properties` map which is lazily initialized. Initialization of this field, done in the private `Configuration.getProps()` method, is rather expensive because it ends up parsing XML configuration files. When cloning a `Configuration`, this `properties` field is cloned if it has been initialized.

In some cases it's possible that `sc.hadoopConfiguration` never ends up computing this `properties` field, leading to performance problems when this configuration is cloned in `SessionState.newHadoopConf()` because each cloned `Configuration` needs to re-parse configuration XML files from disk.

To avoid this problem, we can call `Configuration.size()` to trigger a call to `getProps()`, ensuring that this expensive computation is cached and re-used when cloning configurations.

I discovered this problem while performance profiling the Spark ThriftServer while running a SQL fuzzing workload.

## How was this patch tested?

Examined YourKit profiles before and after my change.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -285,7 +285,18 @@ class SparkContext(config: SparkConf) extends Logging {
...
+  def hadoopConfiguration: Configuration = {
+    // Performance optimization: this dummy call to .size() triggers eager evaluation of
+    // Configuration's internal  `properties` field, guaranteeing that it will be computed and
+    // cached before SessionState.newHadoopConf() uses `sc.hadoopConfiguration` to create
+    // a new per-session Configuration. If `properties` has not been computed by that time
+    // then each newly-created Configuration will perform its own expensive IO and XML
+    // parsing to load configuration defaults and populate its own properties. By ensuring
+    // that we've pre-computed the parent's properties, the child Configuration will simply
+    // clone the parent's properties.
+    _hadoopConfiguration.size()
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: @JoshRosen, out of curiosity, how much did this improve the performance? - just rough estimation is fine. I thought here' in a non-critical path so I thought better readability is prioritized if the gain is rather minor.

Comment (JoshRosen) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: It's a little hard to get a precise end-to-end performance measurement here, unfortunately. In steady state this is't huge, but fixing it cleans up a ton of noise in Java profiler output: after this change I no longer see a bunch of frames from `org.apache.xerces` / `ClassLoader.getResource`, etc.

I submitted this mostly for the sake upstreaming changes from my local fuzz-testing branch (which runs tons of short queries through the ThriftServer).

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Logically, this PR seems to invoke `.size()` every time when we can call `hadoopConfiguration` function. Can we have this invocation at [setupDriverResources](https://github.com/apache/spark/pull/24714/files#diff-364713d7776956cb8b0a771e9b62f82dR514) like the following instead?
```scala
     _hadoopConfiguration = SparkHadoopUtil.get.newConfiguration(_conf)
+    _hadoopConfiguration.size()
```

**Comment by JoshRosen** [comment]
@dongjoon-hyun, good call; updated to move this into `setupDriverResources()`.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thank you, @JoshRosen .

**Comment by HyukjinKwon** [comment]
+1 to me too for the current change

**Comment by hvanhovell** [comment]
@JoshRosen can we just change the initialization order in SparkHadoopUtil? If we set the spark configurations after loading the resource we should get the same result.

**Comment by JoshRosen** [comment]
@hvanhovell, I think that would work, too, and would be broader in scope: the change implemented here was intentionally narrow / surgical in scope, but I think your proposal would also work and should be behaviorally equivalent.

**Pull request closed without merging**